### DISCUSSION TITLE
How to clone a document instance

### DISCUSSION THREAD

**Discussion by saveman71** [Automattic/mongoose#10788] [Q&A] [answered]
Let's say I have a function that takes a Document instance in parameter that _may or may not_ have some populated paths.

I'd like this function to always work a depopulated version of the document, so I could call `.depopulate()` on the document. But this would have the side effect of depopulating the paths for the caller context as well.

How do I clone / deep copy a Document instance to do some work on it without side effects?

**Comment by vkarpov15** [comment, marked as answer]
There are a couple of possibilities that may work depending on your use case.

Easiest way would be to just do `const clonedDoc = new MyModel(...doc.toObject())` or `const clonedDoc = MyModel.hydrate(doc.toObject())`, which gets you a new document with all the same properties as `doc`. You can then depopulate and do whatever you need to.

The downside is that you lose change tracking, so `clonedDoc` wouldn't be aware of what paths in `doc` are modified. But if you don't intend to do `clonedDoc.save()`, this could work.

**Reply by saveman71** [reply]
The use case is to do changelog style tracking, so I'm comparing objects before and after modification, and I don't save them, but also I don't want an "in-between" population generating changes.

So I do a `toObject()` on the depopulated object, which would look like:

```js
track(doc) {
  const cloned = new this._model({...doc.toObject()});
  cloned.depopulate();
  this._original = cloned.toObject({ getters: false, virtuals: false });
}
```

It's a bit weird to call toObject twice (just to depopulate), but I get why it's needed.

An additional option in the toObject() to not include populated paths would fit my use case as well. Is this something that exists / would be welcomed as a feature?

A "clone()" method would make that cleaner but I assume it would essentially be the same logic behind the scenes? (also quite hard to add a clone() method without breaking changes)

**Reply by vkarpov15** [reply]
Re: option to not include populated paths in `toObject()`, we already have that. `const cloned = new this._model(doc.toObject({ depopulate: true }))`. No need to use spread operator, `toObject()` clones `doc`'s data.

It would be similar logic behind the scenes, the difference is that we'd also copy over change tracking info. But if you don't need a fully hydrated copy, you just need a [POJO](https://masteringjs.io/tutorials/fundamentals/pojo) for diffing, you don't need a `clone()` method. `toObject({ depopulate: true })` should be what you need.

**Reply by saveman71** [reply]
Okay, I used the spread operator because you included it in your first answer, so maybe it's best edit it then?

But yes, thanks! I wasn't aware of the depopulate option, it's exactly what I need.

```
track(doc) {
  this._original = cloned.toObject({ getters: false, virtuals: false, depopulate: true });
}
```
